### DISCUSSION TITLE
Why hasOne() is forcing update of related model? How to understand 1-1 relationships?

### DISCUSSION THREAD

**Discussion by aircodepl** [phalcon/cphalcon#15685] [Q&A] [answered]
Hi!

I run today into the same issue: https://github.com/phalcon/cphalcon/issues/12393 on Phalcon 3, tested it right now in Phalcon 5 and I see that it's still the same.

I've created basic models:

```php
<?php

use Phalcon\Mvc\Model;

/** @property Car $car */
class Driver extends Model
{
    public int $id;
    public string $name;
    public int|null $carId;
    public int|null $alive;

    public function initialize()
    {
        $this->useDynamicUpdate(true);

        $this->setSource('drivers');

        $this->hasOne(
            'carId',
            Car::class,
            'id',
            [
                'alias' => 'car'
            ]
        );
    }
}

```

```php
<?php

use Phalcon\Mvc\Model;

class Car extends Model
{
    public int $id;
    public string $name;
    public int|null $alive;

    public function initialize()
    {
        $this->useDynamicUpdate(true);

        $this->setSource('cars');
    }
}
```

Assigned car to Driver and decided to steal it from him later:

```php
 public function indexAction()
    {
        /** @var Driver $driver */
        $driver = Driver::findFirst(
            [
                'conditions' => 'id = :id:',
                'bind' => [
                    'id' => 1
                ]
            ]
        );

        $car = $driver->car;

        echo "Driver name is $driver->name <br>";
        echo "Driver car is $car->name <br>";

        $driver->carId = null;
        $driver->save();

        echo "Ooops! Driver doesn't have car anymore... And there are two cars now.";

        exit;
```

What I would expect is to just set carId to null, so driver doesn't have car anymore, and other driver can take it.

Instead, when I hit `$driver->save();`, Phalcon is trying to create new Car with empty values, which have no sense for me.

**What am I missing?**

A way to avoid this weird (for me) action is to **avoid** use of relationship `$driver->car`. I need to find a car by id instead, which probably creates another SQL query and object in memory, because we use relationship in other files too...

```php
$car = Car::findFirst();
```

The topic that I linked to is suggesting using belongsTo() instead. I don't get why and how I would use it in this example. For me Driver has (or not) one car _(Because in my world Driver cannot have more than one)_. Driver doesn't belong to a Car. And downloading Car instead of using relationship is not what I wanted...

Shouldn't this be fixed? If I want to update/save Car I will call it. For me it seems like a bug or unclear design.

Regards

**Comment by quasipickle** [comment, marked as answer]
Whenever you save a model, relevant related models _should_ be saved.  That's not broken.  I think the issue is how you're trying to disassociate the driver from the car.  I don't know the details, but Phalcon does some magic behind-the-scenes to ensure the relationship connection, and I don't think setting the connecting column to `null` kills that connection.  I think the standard way is to set `car` to `null` like:

```php
$driver->car = null;
$driver->save();
```

**Reply by aircodepl** [reply]
Just tested it. You are right. Thank you.

-----
EDIT:
Actually, after 2-3 weeks later I got this error again for Phalcon 3.*. The only way to remove $driver->car is to NOT use a relation. But at least in Phalcon 5 it seemed to work well : )

**Comment by Mikikou** [comment]
shouldn't your relation be a belongsTo ? because I think so.

hasOne should be used in Car model in your case

$this->hasOne(
            'id',
            Driver::class,
            'carId',
            [
                'alias' => 'driver'
            ]
        );

**Reply by aircodepl** [reply]
Yeah, maybe. I didn't really think of it. It was just a sample code that I wrote quickly just to show the problem ;).
